src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpMetricSerializer.cs

```diff
@@ -0,0 +1,352 @@
+        ReturnMetricListToPool();
+
+        return writePosition;
+    }
+
+    private static void ReturnMetricListToPool()
+    {
+        if (ScopeMetricsList.Count != 0)
+        {
+            foreach (var entry in ScopeMetricsList)
+            {
+                entry.Value.Clear();
+                MetricListPool.Push(entry.Value);
+            }
+
+            ScopeMetricsList.Clear();
+        }
+    }
+
+    private static int WriteResourceMetrics(byte[] buffer, int writePosition, in Resources.Resource? resource, in Dictionary<string, List<Metric>> scopeMetrics)
+    {
+        writePosition = ProtobufOtlpResourceSerializer.WriteResource(buffer, writePosition, resource);
+        writePosition = WriteScopeMetrics(buffer, writePosition, scopeMetrics);
+
+        return writePosition;
+    }
+
+    private static int WriteScopeMetrics(byte[] buffer, int writePosition, in Dictionary<string, List<Metric>> scopeMetrics)
+    {
+        if (scopeMetrics != null)
+        {
+            foreach (KeyValuePair<string, List<Metric>> entry in scopeMetrics)
+            {
+                writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.ResourceMetrics_Scope_Metrics, ProtobufWireType.LEN);
+                int resourceMetricsScopeMetricsLengthPosition = writePosition;
+                writePosition += ReserveSizeForLength;
+
+                writePosition = WriteScopeMetric(buffer, writePosition, entry.Value);
+
+                ProtobufSerializer.WriteReservedLength(buffer, resourceMetricsScopeMetricsLengthPosition, writePosition - (resourceMetricsScopeMetricsLengthPosition + ReserveSizeForLength));
+            }
+        }
+
+        return writePosition;
+    }
+
+    private static int WriteScopeMetric(byte[] buffer, int writePosition, in List<Metric> metrics)
+    {
+        writePosition = ProtobufSerializer.WriteTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.ScopeMetrics_Scope, ProtobufWireType.LEN);
+        int instrumentationScopeLengthPosition = writePosition;
+        writePosition += ReserveSizeForLength;
+
+        Debug.Assert(metrics.Any(), "Metrics collection is not expected to be empty.");
+        var meterName = metrics[0].MeterName;
+        var meterVersion = metrics[0].MeterVersion;
+        var meterTags = metrics[0].MeterTags;
+
+        writePosition = ProtobufSerializer.WriteStringWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.InstrumentationScope_Name, meterName);
+        if (meterVersion != null)
+        {
+            writePosition = ProtobufSerializer.WriteStringWithTag(buffer, writePosition, ProtobufOtlpMetricFieldNumberConstants.InstrumentationScope_Version, meterVersion);
+        }
+
+        if (meterTags != null)
+        {
+            // TODO: DON'T HAVE ANY TESTS THAT EVALUATE THIS
+            foreach (var tag in meterTags)
```

Comment:
`meterTags` is `IEnumerable<T>`? Consider doing a `IReadOnlyList<T>` type check like we do [here](https://github.com/open-telemetry/opentelemetry-dotnet/blob/e3b920c0daaca26ae2326cb00c0a7f758edb7310/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/Serializer/ProtobufOtlpTraceSerializer.cs#L105) to avoid allocating an enumerator. The actual type at runtime _should_ be an `IReadOnlyList<T>`: https://github.com/dotnet/runtime/blob/5d9d6f5cf67ffc74b7224ca4473e6f2992bd75e2/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/Meter.cs#L86